File: benchmarks/livermore/k04_banded.f
C     Kernel 4 -- Banded linear equations
C     From "The Livermore Fortran Kernels" (UCRL-53745), F.H. McMahon, 1986
C
      PROGRAM K04
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      INTEGER REP
      DIMENSION X(1001), Y(1001)
C
C     Initialize arrays using SIGNEL pattern
C
      SCALED = 0.1D0
C
      FUZZ  = 1.2345D-3
      BUZZ  = 1.0D0 + FUZZ
      FIZZ  = 1.1D0 * FUZZ
      DO 10 K = 1, 1001
          BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
          FUZZ = -FUZZ
          Y(K) = (BUZZ - FIZZ) * SCALED
   10 CONTINUE
C
      N = 1001
      M = (1001 - 7) / 2
C
      DO 200 REP = 1, 10000
C         Reset X each rep
          FUZZ  = 1.2345D-3
          BUZZ  = 1.0D0 + FUZZ
          FIZZ  = 1.1D0 * FUZZ
          DO 20 K = 1, 1001
              BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
              FUZZ = -FUZZ
              X(K) = (BUZZ - FIZZ) * SCALED
   20     CONTINUE
C
C         Kernel loop
C         C uses k=6..1000 step m, 0-based; Fortran uses k=7..1001 step m, 1-based
C
          DO 100 K = 7, 1001, M
              LW = K - 6
              TEMP = X(K)
C             C uses j=4..n-1 step 5, 0-based; Fortran uses j=5..n step 5, 1-based
              DO 90 J = 5, N, 5
                  TEMP = TEMP - X(LW) * Y(J)
                  LW = LW + 1
   90         CONTINUE
              X(K) = Y(5) * TEMP
  100     CONTINUE
  200 CONTINUE
C
      WRITE(*,900) X(7)
  900 FORMAT('K4 banded: x(7) = ', E25.15)
      END

File: benchmarks/livermore/build_fast/src/k04_banded.c
/* K4 — Banded linear equations (1-based indexing, matches Fortran exactly)
   Fortran: DIMENSION X(1201), Y(1001), N=1001, M=(1001-7)/2
   DO K=7,1001,M: LW=K-6; DO J=5,N,5: TEMP=TEMP-X(LW)*Y(J); LW=LW+1
   X(K)=Y(5)*TEMP */
#include <stdio.h>
#include <time.h>
#include "signel.h"

#define N     1001
#define NREPS 11250

int main(void) {
    double x[1202], y[1002];

    signel(x+1, 1001);
    signel(y+1, 1001);

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    for (long rep = 0; rep < NREPS; rep++) {
        signel(x+1, 1001);
        long m = (N - 7) / 2;
        for (long k = 7; k <= N; k = k + m) {
            long lw = k - 6;
            double temp = x[k];
            for (long j = 5; j <= N; j = j + 5) {
                temp = temp - x[lw] * y[j];
                lw = lw + 1;
            }
            x[k] = y[5] * temp;
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) * 1e-9;
    printf("elapsed: %.6f s\n", elapsed);
    printf("x[7] = %f\n", x[7]);
    return 0;
}
